examples/seismic/acoustic/wavesolver.py

```diff
@@ -211,10 +219,34 @@ def born(self, dmin, src=None, rec=None, u=None, U=None, m=None, **kwargs):
                          save=False, time_order=self.time_order,
                          space_order=self.space_order, dtype=self.model.dtype)
 
+        kwargs.update({'src': src})
+        kwargs.update({'rec': rec})
+        kwargs.update({'u': u})
+        kwargs.update({'U': U})
+
         # Pick m from model unless explicitly provided
-        if m is None:
-            m = self.model.m
+        kwargs = self.check_input(m, **kwargs)
 
+        kwargs.update({'dm': dmin})
         # Execute operator and return wavefield and receiver data
-        summary = self.op_born.apply(dm=dmin, u=u, U=U, src=src, rec=rec, m=m, **kwargs)
+        summary = self.op_born.apply(**kwargs)
         return rec, u, U, summary
+
+    def check_input(self, m, **kwargs):
```

Comment:
Yeah I don't think this routine is necessary at all. The `wavesolver` should just pass the objects down and let the devito argument engine pick up any changes in the type signature. 

If we want to go further and cause re-compilation, I would propose to improve the operator cashing to take the type signature into account. But as far as I understand it, this is not yet required, and so beyond this PR? @mloubout